doc/developer/design/20230214_transform_using.md

```diff
@@ -0,0 +1,359 @@
+persisting the rest of the ingestion. This MFP fragment will be deferred and applied on the read
+side when a dataflow uses the collection `C`.
+
+With these equations we have accomplished the goal of doing as much work as possible before the
+first persist operator while requiring no past state from `RAW` to resume. To see why that is the
+case, let's describe the state needed by the `Q_expiration` operator. The state of `Q_expiration` at
+time `t` is the set of records `(row, t_record, diff)` for which `t_record <= t && t <
+upper_ts(Q_mfp(row))`. In other words, for every timestamp it needs to remember all past records
+that expire after that timestamp.
+
+If we now look at the output of `Q_safemfp` we see that it converts a `(row, t_record, diff)` update
+to `((row, lower_ts, upper_ts), t_record, diff)` update. Additionally, `Q_expiration` will transform
+each `((row, lower_ts, upper_ts), t_record, diff)` update into two:
+
+1. `((row, lower_ts, upper_ts), t_record, diff)`
+2. `((row, lower_ts, upper_ts), upper_ts, -diff)`
+
+Given the above we can see that at the beginning of an ingestion execution `Q_expiration` can
+recover its in-memory state at time `t` by taking a snapshot of `C_intermediate` (not `RAW`) at time
+`t`, which is available because it is stored in persist! This is because any record that may expire
+in the future is maintained until the end of its validity period and `Q_mfp` has preserved the
+previously computed timestamps. This technique is similar to how the `UPSERT` envelope recovers its
+state from its output.
+
+### Planner/Optimizer stabilization
+
+When accepting a `CREATE SOURCE` statement with a `TRANSFORM USING` clause we are at that moment
+making a promise that that transformation will always be compatible with the storage execution
+model.
+
+In practice, since we have limited the accepted queries to only those that fully compile to an MFP,
+this means that the planner/optimizer must maintain the following invariant:
+
+> If a query `Q` fully optimizes to an MFP at some version `n` of Materialize, then it will continue
+> to plan and optimize to an MFP for all versions `n' >= n`. The MFPs are allowed to vary across
+> versions to allow for new optimizations to be taken advantage of.
+
+## Future improvements
+
+The feature as described includes a lot of limitations that are necessary with our current appetite
+for stability promises. Here we explore how we might lift some of them in the future.
+
+### Accept `TRANSFORM USING` for non-append only collections
+
+The fisrt limitation that we can lift is allowing transforming non-append only streams. This would
+allow users to transform the output of envelopes other than NONE, and also to transform the output
+of postgres sources.
+
+In order to do that we'll have to address the sources of non-determinism in the evaluation semantics
+of a SQL query. Here are a few thoughts on how each could be addressed but more design work is
+needed to fully spec these out.
```

Comment:
Looks like you wanted to move the next section behind the "thoughts on removing non-determinism" sections?